File: Tool/ImageConversion/Source/Internal/Memory.c
#include "Memory.h"



typedef UInt64 AlignType;

static const SizeType AlignSize = sizeof(AlignType);


static Memory Construct(SizeType byteCount)
{
	Memory ret;
	ret.buffer		= FAllocator.Allocate(byteCount);
	ret.byteCount	= byteCount;

	return ret;
}

static Void Destruct(Memory* instance)
{
	if ( !instance )
		return;

	FAllocator.Release(instance->buffer);
}

static Void Clear(Void* buffer, SizeType byteCount)
{
	SizeType count	= byteCount / AlignSize;
	SizeType mod	= byteCount % AlignSize;
	SizeType offset = byteCount - mod;

	for (SizeType index = 0; index < count; index++)
		CAST(AlignType*, buffer)[index] = 0;

	for (SizeType index = 0; index < mod; index++)
		CAST(Byte*, buffer)[offset + index] = 0;
}

static Void Copy(const Void* from, Void* to, SizeType byteCount)
{
	SizeType count	= byteCount / AlignSize;
	SizeType mod	= byteCount % AlignSize;
	SizeType offset	= byteCount - mod;

	for (SizeType index = 0; index < count; index++)
		CAST(AlignType*, to)[index] = CAST(const AlignType*, from)[index];

	for (SizeType index = 0; index < mod; index++)
		CAST(Byte*, to)[offset + index] = CAST(const Byte*, from)[offset + index];
}

static Bool Equal(const Void* bufferA, const Void* bufferB, SizeType byteCount)
{
	SizeType count	= byteCount / AlignSize;
	SizeType mod	= byteCount % AlignSize;
	SizeType offset = byteCount - mod;

	for (SizeType index = 0; index < count; index++)
		if (
			CAST(const AlignType*, bufferA)[index] !=
			CAST(const AlignType*, bufferB)[index]
			)
			return False;

	for (SizeType index = 0; index < mod; index++)
		if (
			CAST(const Byte*, bufferA)[offset + index] !=
			CAST(const Byte*, bufferB)[offset + index]
			)
			return False;

	return True;
}


struct FMemory FMemory =
{
	Construct,
	Destruct,
	Clear,
	Copy,
	Equal,
};

File: CSharp/Engine/Runtime/Core/System/Memory.cs



namespace Engine
{
	using AlignType = System.UInt64;


	public class Memory
	{
		public const int AlignSize = sizeof(AlignType);


		public unsafe static void Clear(void* pointer, int byteCount)
		{
			int count	= byteCount / AlignSize;
			int mod		= byteCount % AlignSize;
			int offset	= byteCount - mod;

			for ( int index = 0; index < count; index++ )
				( (AlignType*)pointer )[index] = 0;

			for ( int index = 0; index < mod; index++ )
				( (byte*)pointer )[ offset + index ] = 0;
		}

		public unsafe static void Copy(void* from, void* to, int byteCount)
		{
			int count	= byteCount / AlignSize;
			int mod		= byteCount % AlignSize;
			int offset	= byteCount - mod;

			for ( int index = 0; index < count; index++ )
				( (AlignType*)to )[index] = ( (AlignType*)from )[index];

			for ( int index = 0; index < mod; index++ )
				( (byte*)to )[ offset + index ] = ( (byte*)from )[ offset + index ];
		}

		public unsafe static bool Equal(void* pointerA, void* pointerB, int byteCount)
		{
			int count	= byteCount / AlignSize;
			int mod		= byteCount % AlignSize;
			int offset	= byteCount - mod;

			for ( int index = 0; index < count; index++ )
				if (
					( (AlignType*)pointerA )[index] != ( (AlignType*)pointerB )[index]
					)
					return false;

			for ( int index = 0; index < mod; index++ )
				if (
					( (byte*)pointerA )[ offset + index ] != ( (byte*)pointerB )[ offset + index ]
					)
					return false;

			return true;
		}
	}
}